sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -1073,8 +1135,51 @@ private[joins] object LongHashedRelation {
         return HashedRelationWithAllNullKeys
       }
     }
-    map.optimize()
-    new LongHashedRelation(numFields, map)
+
+    val reorderMap = reorderFactor.exists(_ * map.numUniqueKeys <= map.numTotalValues)
+    val finalMap = if (reorderMap) {
+      // reorganize the hash map so that nodes of a given linked list are next to each other in
+      // memory.
+      logInfo(s"Reordering LongToUnsafeRowMap, numUniqueKeys: ${map.numUniqueKeys}, " +
+        s"numTotalValues: ${map.numTotalValues}")
+      // An exception due to insufficient memory can occur either during initialization or while
+      // adding rows to the map.
+      // 1. Failure occurs during initialization i.e. in LongToUnsafeRowMap.init:
+      // release of the partially allocated memory is already taken care of in the
+      // LongToUnsafeRowMap.ensureAcquireMemory method thus no further action is required.
+      // 2. Failure occurs while adding rows to the map: the partially allocated memory
+      // is not cleaned up, thus LongToUnsafeRowMap.free is invoked in the catch clause.
+      var maybeCompactMap: Option[LongToUnsafeRowMap] = None
+      try {
+        maybeCompactMap = Some(new LongToUnsafeRowMap(taskMemoryManager,
+          Math.toIntExact(map.numUniqueKeys)))
+        val compactMap = maybeCompactMap.get
+        val resultRow = new UnsafeRow(numFields)
+        map.keys().foreach { rowKey =>
+          val key = rowKey.getLong(0)
+          map.get(key, resultRow).foreach { row =>
+            compactMap.append(key, row)
+          }
```

Comment:
[question] IIUC, at this point of time we will have both the old map as well as the new compact map, in memory and we would be holding 2X memory (at this time) which we were holding before (as we just had 1 map). This itself can cause us OOM, do we need some heuristic here ? or perhap's delete / free (not sure if at present we have this functionality) enteries once they have been entered in the  compactMap ?

Reply:
In case an OOM-related exception is thrown while appending to the map, we invoke `maybeCompactMap.foreach(_.free())` in the catch clause which releases the memory of the `compactMap`.

https://github.com/apache/spark/pull/35047/files#diff-127291a0287f790755be5473765ea03eb65f8b58b9ec0760955f124e21e3452fR1171

Reply:
What I meant here is if we know before hand we don't have this much memory available we should try to short circuit this re-write by not attempting to do do ? we have this `getTotalMemoryConsumption` from earlier map, and we can calculate available memory to us during start, if we know we can't have / support 2X , maybe we will save time avoiding the extra work if the exception happens.

Reply:
That sounds like a good idea, however, `TaskMemoryManager` does not expose API to fetch available execution memory. There does exist `MemoryManager.getExecutionMemoryUsageForTask` but unfortunately `MemoryManager` is not accessible at the place of building the Map.
https://github.com/apache/spark/blob/4c806728eb955c07f7e095a6ff085a50d2eef806/core/src/main/scala/org/apache/spark/memory/MemoryManager.scala#L216

Furthermore, re-building the map takes a fraction amount of the time of the entire query execution.
I observed the following times while running a test with this feature:

<table>
   <body>
       <tr>
         <td>Stream side</td>
         <td>300M rows</td>
      </tr>
      <tr>
         <td>Build side</td>
         <td>90M rows</td>
      </tr>
      <tr>
         <td>Rebuilding the map</td>
         <td>4 seconds (diff from the logs)</td>
      </tr>
      <tr>
         <td>Total query execution time with optimization enabled</td>
         <td>3.9 minutes</td>
      </tr>
      <tr>
         <td>Total query execution time with optimization disabled</td>
         <td>9.2 minutes</td>
      </tr>
   </tbody>
</table>

Reply:
> It's a valid concern that this puts more memory pressure on the driver. Is it possible to improve the relation-building logic and make it co-locate the values of the same key? Then we don't need to rewrite the relation.

+1 on this. Beside memory concern, I am also not sure of the performance penalty when doing shuffled hash join with large hash table in memory. We need to probe each key in the hash table to rebuild the table, and we kind of waste of the time of 1st build of table.

Reply:
> Is it possible to improve the relation-building logic and make it co-locate the values of the same key?

Since there is no ordering guarantee with `Iterator[InternalRow]`, I don't believe there is a way to co-locate the values of the same key when building the relation, but let me think about it again.

Reply:
> I am also not sure of the performance penalty when doing shuffled hash join with large hash table in memory. We need to probe each key in the hash table to rebuild the table, and we kind of waste of the time of 1st build of table.

I totally agree regarding wasting time on 1st build of the table. However, if we consider the amount of time taken to re-build the table, it is a minuscule amount when compared to the overall execution time of the query.
It can be clearly seen in the above example.
Also, in the above case, I was running the spark application in local mode. If we run the application on Yarn or K8s, the overall execution time will increase due to added network cost and the table rebuild time would further become a smaller fraction of the overall time.